### ISSUE TITLE
Testing with P100 on Kaggle

### ISSUE THREAD

**Issue by Iambestfeed** [OpenLMLab/LOMO#13]
I have a question, is it possible to train a GPT2 model with LOMO?
And it seems that DeepSpeed ​​is required to use LOMO right? (what I wonder is even if I only have 1 GPU 4090 will DeepSpeed still be effective?)

**Comment by KaiLv69** [comment]
> I have a question, is it possible to train a GPT2 model with LOMO?

Sure. LOMO is model-agnostic.

> And it seems that DeepSpeed ​​is required to use LOMO right? (what I wonder is even if I only have 1 GPU 4090 will DeepSpeed still be effective?)

We use DeepSpeed (ZeRO-3) for parameters partition. DeepSpeed is not necessary (but works well) when model parallelism is not required. FYI: https://github.com/OpenLMLab/LOMO/blob/548d8e0af5ce73c467c39c7e1471f3ba6b8b8785/src/lomo.py#L46

DeepSpeed also works well on 1 GPU. Parameters will still be registered `ds_tensor` and some other attributes, but will not be partitioned.

**Comment by Iambestfeed** [comment]
@KaiLv69 Thanks for the response. 
I have tested LOMO with BLOOM 560m and have some comments as follows:
1. It seems that batchsize doesn't really matter (in some cases I find bs = 16 has better loss than bs = 64)
2. Is the application of lora init and then Lomo training really effective? What if the process is reversed?
3. It seems that the model will have a period of instability in the first epoch at the start of training, is there any reason for the loss to become so uncontrollably large?
![image](https://github.com/OpenLMLab/LOMO/assets/95571916/f5c94618-050e-4dfe-9026-cd564bd4bd3e)

**Comment by KaiLv69** [comment]
> 1. It seems that batchsize doesn't really matter (in some cases I find bs = 16 has better loss than bs = 64)

Got it! Thanks for your feedback.

> 2. Is the application of lora init and then Lomo training really effective? What if the process is reversed?

The order of initializing LoRA and LOMO is not important because we skip the parameters of LoRA when initializing LOMO. As for the effectiveness, we think LoRA and LOMO tune models in different ways and the performance may depend on the feature of downstream tasks.

> 3. It seems that the model will have a period of instability in the first epoch at the start of training, is there any reason for the loss to become so uncontrollably large?

I think maybe it's due to the model size. As mentioned in our paper, we suggest LOMO may be a practical optimizer for **large** language models.

<img width="1816" alt="image" src="https://github.com/OpenLMLab/LOMO/assets/39761308/50fe03c9-a74c-41a2-8956-94a89ad26602">

**Comment by Iambestfeed** [comment]
Oh, my bad.
I tested on BLOOM 560m so it doesn't seem like a good choice to test some properties in your study. Thanks for clarifying this for me.
I plan to apply **Qlora** in combination with Lomo. What size model do you think I should be testing on that would be suitable for these tests (the end goal of this is I try to improve the finetune process with our resource shortage) .

**Comment by KaiLv69** [comment]
Smaller models (e.g. BLOOM 560m) are good choices for debugging. To improve performance, I think the larger the better. You can choose the largest model that can fit into your resource.

**Comment by Iambestfeed** [comment]
@KaiLv69 
Maybe this annoys you, but I read something from Qlora issues. It seems that Qlora is not really suitable for DeepSpeed.
And my question is is there any way for me to implement LOMO without DeepSpeed support? (May be I need to deploy Lomo testing on Kaggle - a place for me to test many things except DeepSpeed <("))

**Comment by KaiLv69** [comment]
Glad to help you. We have indeed implemented the version without DeepSpeed. As said here, LOMO will detect whether DeepSpeed has been used to register different hook functions.

> We use DeepSpeed (ZeRO-3) for parameters partition. DeepSpeed is not necessary (but works well) when model parallelism is not required. FYI: https://github.com/OpenLMLab/LOMO/blob/548d8e0af5ce73c467c39c7e1471f3ba6b8b8785/src/lomo.py#LL43-L46
